packages/react-dom/src/events/ReactBrowserEventEmitter.js

```diff
@@ -84,22 +85,47 @@ import isEventSupported from './isEventSupported';
  *    React Core     .  General Purpose Event Plugin System
  */
 
-const alreadyListeningTo = {};
-let reactTopListenersCounter = 0;
+const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
+const elementListeningObjects:
+  | WeakMap
+  | Map<
+      Document | Element | Node,
+      ElementListeningObject,
+    > = new PossiblyWeakMap();
```

Comment:
Ahh, :fire:. This is a great update!

Reply:
I've been thinking about something similar for wrapper state meta data that is attached to inputs, like here:

https://github.com/facebook/react/blob/d0289c7e3a2dfc349dcce7f9eb3dee22464e97bd/packages/react-dom/src/client/ReactDOMInput.js#L118-L126